Fixes #27067 - IPAM Integration with phpIPAM

grizzthedj opened the pull request:
This PR is a patch to support of the use of plugins for external IPAM use, with opensource provider phpIPAM(https://phpipam.net). The primary use case is to provide the next available IPv4 address from phpIPAM, for a given phpIPAM subnet. 

The functionality that Foreman will use for external IPAM is encompassed in the below plugins(which are both new).

**foreman_ipam(v0.0.7)**
**Git:** https://github.com/grizzthedj/foreman_ipam
**Gem:** https://rubygems.org/gems/foreman_ipam

**smart_proxy_ipam(v0.0.13)**
**Git:** https://github.com/grizzthedj/smart_proxy_ipam
**Gem:** https://rubygems.org/gems/smart_proxy_ipam

The `foreman_ipam` plugin provides a very basic dashboard to display sections and subnets from phpIPAM, and the `smart_proxy_ipam` plugin communicates with the phpIPAM API.

Features and functionality

* Simple IPAM Dashboard page(Infrastructure => IPAM Dashboard) to view sections and subnets for phpIPAM. More features will be added here based on internal user feedback and from the community.
* When creating a phpIPAM subnet(IPAM type = External IPAM) in Foreman, subnet/mask must exist in phpIPAM, or error will be thrown
* When creating a new host in Foreman, auto suggested IP's from phpIPAM will be persisted back to phpIPAM on Host submit.
* When updating the IP address on a NIC that uses a phpIPAM subnet, the old IP will be deleted from phpIPAM, and new one will be added.
* Deleting a host in Foreman will delete all NIC ip's that used a phpIPAM subnet
* Deleting a NIC from a host will also delete the IP from phpIPAM
* Deleting a subnet in Foreman will NOT delete the associated subnet in phpIPAM. This could easily be added later if it makes sense to the community, however initial thoughts are to not delete it. 

NOTE: When creating a host in Foreman and adding multiple nics/interfaces, the same IP will be suggested from phpIPAM since it is not actually reserved in phpIPAM until you submit the host. The workaround for multiple nics is to use the suggested ip for the first interface(e.g. 10.0.0.1), then manually increment the IP for the second interface(e.g. 10.0.0.2). Errors will be thrown however if IP's are already taken.

Prerequisites for testing:
* Working Foreman environment.
* A phpIPAM instance that the Forman Smart Proxy has network connectivity to.
* The API must be enabled in the phpIPAM settings, and `Prettify Links` also enabled.
* An API user in phpIPAM must be created. The credentials for this user are to be put in the `config/settings.d/external_ipam.yml` file in the Smart Proxy core

Commit:
Fixes #27067 - External IPAM Integration with phpIPAM

tbrisker commented:
ok to test

lzap commented on app/controllers/application_controller.rb:
```diff
@@ -71,6 +73,13 @@ def authorize
+
   def deny_access
     (User.current.logged? || request.xhr?) ? render_403 : require_login
   end
```
We cannot extend IPAM modes like that, this is actually done in the runtime, every request would add new entry there.

You defined a nice Smart Proxy API, just add the "External IPAM" into IPAM::MODES constant. I suggest it to be `:external => N_("External IPAM")`. No need of doing this if a plugin is active or not. Your plugin can add extra features like the widget for phpMyIpam or other nice to have things, but the core should be in the main Rails app.

lzap commented on app/models/host/managed.rb:
```diff
@@ -871,6 +873,37 @@ def validate_dns_name_uniqueness
+      end
+    end
+  end
+
```
So we have an orchestration "framework" in Foreman that you should use. Search for `orchestration/dhcp.rb` for an example, you should create new file `ipam.rb` there and queue create, update and delete methods the same way as dhcp does. It does support transaction compensation - when something goes wrong it performs a rollback.

lzap commented on app/models/nic/base.rb:
```diff
@@ -108,6 +110,18 @@ def self.allowed_types
+        end
+      end
+    end
+
```
The orchestration framework can be used for both hosts and NICs.

lzap commented on app/models/subnet.rb:
```diff
@@ -58,6 +59,12 @@ def model_name
+    :feature => N_('external_ipam'),
+    :label => N_('IP Address Management with various external IPAM providers'),
+    :api_description => N_('IPAM Proxy ID to use within this subnet'),
+    :description => N_('IPAM Proxy to use within this subnet')
```
I suggest to name the feature as `External IPAM`.

grizzthedj replied:
How do I best handle feature names with spaces? In the Smart Proxy plugin, I have changed the plugin name as follows: 

```
  class Plugin < ::Proxy::Plugin
    plugin 'External IPAM', Proxy::Ipam::VERSION

    http_rackup_path File.expand_path('ipam_http_config.ru', File.expand_path('../', __FILE__))
    https_rackup_path File.expand_path('ipam_http_config.ru', File.expand_path('../', __FILE__))
  end
```

The Smart Proxy does not pick this up on startup when there is a space in the name. The Smart proxy `associate_features` method is not able to map the feature names it seems due to the space in the name. 

Update: Also seems that you cannot map the IPAM proxy to subnets

lzap replied:
Oh right, then I suggest to use only "IPAM" for the Smart Proxy API and "External IPAM" for the Foreman core IPAM provider.

grizzthedj replied:
@lzap I'm having a tough time getting this to work using "IPAM" for the Smart Proxy API and "External IPAM" for the Foreman core IPAM provider. Can we name it `IPAM` in both places?

lzap commented on app/models/subnet.rb:
```diff
@@ -15,6 +15,7 @@ class Subnet < ApplicationRecord
   include Parameterizable::ByIdName
   include Exportable
   include BelongsToProxies
+  include Foreman::Renderer::Scope::Macros::Base
```
Why you add it here and pretty much everywhere else?

lzap commented on app/models/subnet.rb:
```diff
@@ -218,6 +234,19 @@ def unused_ip(mac = nil, excluded_ips = [])
+        self.errors.add :network, _('Subnet not found in the configured phpIPAM instance')
+      end
+    end
+  end
```
Please remove all `plugin_present?` calls, the External IPAM feature should work without your plugin - users will be able to associate External IPAM only if there is a smart proxy with that feature reported. It's fine to allow them to set External IPAM for a Subnet, just raise an error if user attemts to create host in subnet with external IPAM but feature will not be present.

lzap commented on app/models/subnet/ipv4.rb:
```diff
@@ -13,6 +13,8 @@ class Subnet::Ipv4 < Subnet
 
   before_validation :cleanup_addresses
 
+  extend Foreman::Renderer::Scope::Macros::Base
```
These must be all removed.

lzap commented on app/models/subnet/ipv4.rb:
```diff
@@ -26,7 +28,9 @@ def validate_ip(ip)
-    [:dhcp, :db, :random_db, :none]
+    modes = [:dhcp, :db, :random_db, :none]
+    modes.unshift(:external_ipam) if plugin_present?('foreman_ipam')
+    modes
```
External IPAM can be hardcoded.

lzap commented on db/migrate/20190617165412_rename_phpipam_id_to_external_ipam_id_in_subnets.rb:
```diff
@@ -0,0 +1,9 @@
+  def down
+    rename_column :subnets, :external_ipam_id, :phpipam_id
+  end
+end
```
Please squash the two migrations into one, just add "external_ipam_id" column to the subnet.

lzap commented on lib/proxy_api/external_ipam.rb:
```diff
@@ -0,0 +1,127 @@
+module ProxyAPI
+  class ExternalIpam < ProxyAPI::Resource
+    def initialize(args)
+      @url = args[:url] + "/phpipam"
```
The smart proxy endpoint must be just `/ipam` - a generic one.

lzap commented on lib/proxy_api/external_ipam.rb:
```diff
@@ -0,0 +1,127 @@
+      get("/next_ip?cidr=#{cidr}")
+    rescue => e
+      raise ProxyException.new(url, e, N_("Unable to retrieve the next available IP for subnet %s from phpIPAM."), cidr)
+    end
```
When users create many hosts concurrently, this can easily run into race condition - a single IP is returned for multiple hosts. I suggest to put MAC as a required parameter and the implementation should memorize what IP it offered to which MAC addresses and never offer the same IP for different MACs. We actually do this in the subnet API in smart proxy - offered IPs are memorized for few minutes. I suggest you take the same approach.

Also document how subnet input param should look like - is that CIDR address?

lzap commented on lib/proxy_api/external_ipam.rb:
```diff
@@ -0,0 +1,127 @@
+      get("/sections")
+    rescue => e
+      raise ProxyException.new(url, e, N_("Unable to obtain sections from phpIPAM."))
+    end
```
Document the return value, what are permissions? Strict mode? I suggest you only return list of names, this must be generic so other implementations can do similar, they might not have sections but something else. If name is not unique in the database, then return just id + name.

Also rename phpIPAM to IPAM this must be generic.

lzap commented on lib/proxy_api/external_ipam.rb:
```diff
@@ -0,0 +1,127 @@
+      get("/sections/#{section_id}/subnets")
+    rescue => e
+      raise ProxyException.new(url, e, N_("Unable to obtain subnets for section_id %s from phpIPAM."), section_id.to_s)
+    end
```
Consider using just section name instead of id if its unique. Also return subnets only in simplest format, no extra fields only those which are really necessary. Make this a generic API.

grizzthedj replied:
We could use section name(it is unique), however this would mean an extra API call to get the id of the section, which in turn would be used to get the subnets collection. Unfortunately the phpIPAM API doesn't support getting the subnets be section name, you can only obtain with section_id.

lzap replied:
For the sake of clean generic external IPAM API I think it is probably worth it. However I don't insist.

What you should probably do is to design the Smart Proxy API in a way that sections are optional - this call should also work without any section provided. In that case, your implementation should probably raise an error "section is required" or something, but other implementation should go through.

lzap commented on lib/proxy_api/external_ipam.rb:
```diff
@@ -0,0 +1,127 @@
+      get("/get_subnet?cidr=#{cidr}")
+    rescue => e
+      raise ProxyException.new(url, e, N_("Unable to obtain subnet %s from phpIPAM."), cidr)
+    end
```
Consider to return only relevant data, document all the fiels returned please.

lzap commented on lib/proxy_api/external_ipam.rb:
```diff
@@ -0,0 +1,127 @@
+      parse(post({:ip => ip, :cidr => cidr}, "/delete_ip_from_subnet"))
+    rescue => e
+      raise ProxyException.new(url, e, N_("Unable to delete IP %s from the subnet %s in phpIPAM."), ip, cidr)
+    end
```
Here and above - document how subnet input param should look like. I disencourage from using database IDs because other implementations might not even use RDBM.

lzap commented on lib/proxy_api/external_ipam.rb:
```diff
@@ -0,0 +1,127 @@
+      response = Net::HTTP.start(uri.hostname, uri.port) {|http| http.request(request)}
+      JSON.parse(response.body)
+    end
+  end
```
If you can improve our current method, do it but take special care - tests must pass, this is indeed not the best part of our codebase :-)

grizzthedj replied:
Perhaps I will take a look, but will be in a different PR

lzap commented on locale/ca/foreman.po:
```diff
@@ -9931,7 +9931,7 @@ msgstr "Podeu afegir claus públiques SSH a un usuari a Foreman."
 msgstr "Podeu trobar «The Foreman» a la xarxa de %{freenode} (irc.freenode.net).  Per al suport general, si us plau, visiteu #theforeman i per al xat específic de desenvolupament, si us plau, visiteu #theforeman-dev."
 
-msgid "You can select one of the IPAM modes supported by the selected IP protocol:<br/><ul><li><strong>DHCP</strong> - will manage the IP on DHCP through assigned DHCP proxy, auto-suggested IPs come from DHCP <em>(IPv4)</em></li><li><strong>Internal DB</strong> - use internal DB to auto-suggest free IP based on other interfaces on same subnet respecting range if specified, useful mainly with stat ...
+msgid "You can select one of the IPAM modes supported by the selected IP protocol:<br/><ul><li><strong>DHCP</strong> - will manage the IP on DHCP through assigned DHCP proxy, auto-suggested IPs come from DHCP <em>(IPv4)</em></li><li><strong>Internal DB</strong> - use internal DB to auto-suggest free IP based on other interfaces on same subnet respecting range if specified, useful mainly with stat ...
```
Please drop all `locale/` changes from this patch, we do gettext extractions separately on a regular basis.

lzap commented on test/factories/subnet.rb:
```diff
@@ -9,6 +9,11 @@
+    after(:build) do |subnet|
+      subnet.class.skip_callback(:validation, :after, :validate_against_external_ipam, raise: false)
+    end
+
```
Once you refactor your code to use orchestration, then write tests similarly as we have for DHCP, TFTP and other orchestration.

lzap reviewed (changes requested):
Very nice patch, it's shaping up. Overall the direction is good, it needs many changes tho.

My idea was/is to create new External IPAM feature that will be available in Foreman core for everybody. You correctly added smart proxy feature, add also IPAM MODE, you do not need to do this if plugin is present or not - the entry can be there just like that. Make sure error is thrown when user attempts to set it when External IPAM is not associated, or when a host is created/edited/deleted and things are not set up properly.

Then instead doing your own validation hooks, use our orchestration approach. It allows you to queue action items and that queue is processed by our framework one by one. You will see many debug messages in Rails log.

The code needs more Ruby or ActiveRecord improvements but first let's attack those big things. Good stuff!

grizzthedj commented:
Pushed most of the request changes - continuing to work on the others that have not been marked as resolved.

lzap commented on app/models/concerns/orchestration/external_ipam.rb:
```diff
@@ -0,0 +1,90 @@
+  def external_ipam?
+    # host.managed? and managed? should always come first so that orchestration doesn't
+    # even get tested for such objects
+    subnet&.ipam&.downcase == "external ipam"
```
Probably a helper method in subnet would be better.

lzap commented on app/models/concerns/orchestration/external_ipam.rb:
```diff
@@ -0,0 +1,90 @@
+    return false if new_record?
+    old = find_nic(id)
+    old['ip'] != ip
+  end
```
Why you search for nic, this module is mixed into NIC::Managed so to access the id just cell `self.ip` and `old.ip`.

lzap reviewed (commented):
Good changes, going the right direction. Keep it up!

grizzthedj commented:
Added tests and moved `external_ipam?` method to helper. 

@lzap Did you see my comment related to renaming the feature as `External IPAM`? I'm having a tough time getting this to work using "IPAM" for the Smart Proxy API and "External IPAM" for the Foreman core IPAM provider, and was wondering if we can just name it `IPAM` everywhere, or leave it as is.

grizzthedj commented:
@lzap - Pushed changes to add in memory cache to hold ip and mac address in the case of a race condition when multiple hosts are being provisioned.

lzap commented:
Sorry for the long delay caused by vacations. To shut the bot, just squash all your commits into one starting with "Fixes #27067" and then everytime you add new one just start the message with "Refs #27067". Let me take a quick look.

lzap commented on app/models/concerns/orchestration/external_ipam.rb:
```diff
@@ -0,0 +1,79 @@
+      subnet.external_ipam_proxy.add_ip_to_subnet(ip, subnet)
+    else
+      self.errors.add :ip, :taken
+      self.errors.add :interfaces, _('Some interfaces are invalid')
```
This is a terrible error we do, I would love to get rid of it in the future.

Please report some nice error like:

          self.errors.add :ip, _('External IPAM does not have available IP')

ekohl commented:
You can also edit the description to start with `[WIP] ` which also signals to reviewers what to expect. Do remember to remove it, then push your final version to get the bot to run again.

lzap commented on app/services/ipam/external_ipam_service.rb:
```diff
@@ -0,0 +1,115 @@
+    # TODO: Make this configurable?
+    MAX_RETRIES = 5
+    # The delay to wait in seconds before clearing the in memory cache
+    CLEAR_CACHE_DELAY = 180
```
I am afraid this approach will not work. Foreman uses multiple processes, this means that every process will get its own cache = race conditions.

What I was initially suggesting was to take the similar approach which our DHCP smart proxy module does in `unused_ip` call: it "remembers" list of pre-allocated IP and if there is another call requesting new suggested IP it will not use it. There is a helper class FreeIps in the DHCP common module which you can use in your Smart Proxy module (just make dhcp_common as a dependency of your module and you can use it). Since Smart Proxy is a multi-threaded app (single process), there is a mutex that protects concurrent access and this works fine. There is a background thread which cleans up allocated IP addresses every X minutes (configurable interval).

It should be pretty easy to implement this on the Smart Proxy, here in Foreman core we do not want any of this code - here it should be just simple HTTP(s) call. Apologies if this was not clear from the beginning, my fault.

grizzthedj replied:
Thanks @lzap - I will take a look at the DHCP implementation.

lzap commented on app/views/subnets/_fields.html.erb:
```diff
@@ -8,13 +8,14 @@
                                    "<li><strong>Internal DB</strong> - use internal DB to auto-suggest free IP based on other interfaces on same subnet respecting range if specified, useful mainly with static boot mode <em>(IPv4, IPv6)</em>, preserves natural ordering</li>" +
                                    "<li><strong>Random DB</strong> - same as Internal DB but randomizes results to prevent race conditions <em>(IPv4)</em></li>" +
                                    "<li><strong>EUI-64</strong> - will assign the IPv6 address based on the MAC address of the interface <em>(IPv6)</em></li>" +
+                                   "<li><strong>External IPAM</strong> - will auto-suggest the next available IPv4 address based on the selected IPAM provider. <em>NOTE: foreman_ipam plugin must be enabled to use this option, and phpIPAM is the only supported provider</em></li>" +
```
Please do this: "<li><strong>External IPAM</strong> - will auto-suggest the next available address via an external IPAM Smart-proxy plugin <em>(IPv4)</em></li>" +

lzap commented on lib/proxy_api/external_ipam.rb:
```diff
@@ -0,0 +1,140 @@
+      get("/sections")
+    rescue => e
+      raise ProxyException.new(url, e, N_("Unable to obtain sections from External IPAM."))
+    end
```
Still not sold about database "IDs", I assume "name" is unique in MyIPAM then use that to find sections. There will be implementations of IPAM without database I am pretty sure.

lzap commented on lib/proxy_api/external_ipam.rb:
```diff
@@ -0,0 +1,140 @@
+    private
+
+    # TODO: Use ProxyAPI::Resource.parse instead for GET(for consistency), once it has been
+    #       refactored to handle error messages better
```
Feel free now, looks good.

lzap reviewed (changes requested):
Ok it is looking good, you can now squash and fix tests after required changes. Great work here.

grizzthedj commented:
@lzap - I have moved the cache code out of Foreman into the smart proxy plugin. Seems to be working well. I did not use the dhcp_common module as it seems this only caches the ip address only. For our use cases, it is possible for the same IP's to exist in different subnets. Plus we also want to use mac address as a key, so that we don't suggest the same IP for different mac addresses. 

You can see the implementation here: 

https://github.com/grizzthedj/smart_proxy_ipam/blob/master/lib/smart_proxy_ipam/phpipam/phpipam_client.rb

I have also updated the `get_subnets` endpoint to accept `section name` as the parameter, not `id`.

From a testing perspective, I took a stab at it, but there are 6 failing tests in this branch(below). **The exact same 6 tests are also failing on the `develop` branch when I run the test suite locally.** Perhaps something on my MAC is not configured, or something is misconfigured? 

I also noticed that there are 195 failures in the `foreman` tests below. What tests are these? Is there a way to run them locally before pushing?

```
37789 tests run, 15 skipped, 195 failed.
```

Welcome any feedback.

Failing tests:
```
1). Failure: 
VariableLookupKeyTest#test_0005_should not create smart variable with invalid default_value [foreman/test/models/lookup_keys/variable_lookup_key_test.rb:90]:
Validation succeeded for create with invalid default_value: variable_type: yaml default_value: {NFoJuBYnMk:EjsTqqalxa})

2). Failure: 
AccessPermissionsTest#test_0780_route api/v2/fake_with_filename/index should have a permission that grants access [foreman/test/unit/shared/access_permissions_test_base.rb:29]:
permission for api/v2/fake_with_filename/index not found, check access_permissions.rb.
Expected [] to not be empty.

3). Failure: 
Api::V2::RolesControllerTest#test_0013_org admin should not create roles by default [foreman/test/controllers/api/v2/roles_controller_test.rb:252]:
--- expected
+++ actual
@@ -1 +1 @@
-"Missing one of the required permissions: create_roles"
+"Missing one of the required permissions: create_roles, create_roles"

4). Failure: 
LookupKeyTest#test_0023_should not update with invalid parameter types [foreman/test/models/lookup_key_test.rb:448]:
Can update lookup key with invalid data {:sc_type=>"yaml", :value=>"{a:test}"}

5). Failure:
Api::V2::LocationsControllerTest#test_0036_org admin should not create locations by default [foreman/test/controllers/api/v2/locations_controller_test.rb:407]:
--- expected
+++ actual
@@ -1 +1 @@
-"Missing one of the required permissions: create_locations"
+"Missing one of the required permissions: create_locations, create_locations"

6). Failure:
TemplateRenderJobTest::processing#test_0002_render report and delivers it to mail [foreman/app/jobs/template_render_job.rb:13]:
unexpected invocation: ReportMailer.report({"foo" => "bar", "send_mail" => true, "mail_to" => "this@email.cz", "gzip" => true}, "result", {:start => 2019-08-22 12:35:16 -0400 (1566491716.613018 secs), :end => 2019-08-22 12:35:16 -0400 (1566491716.613051 secs)})
unsatisfied expectations:
- expected exactly once, not yet invoked: #<Mock:mailer>.deliver_now(any_parameters)
- expected exactly once, not yet invoked: #<AnyInstance:ReportComposer>.report_filename(any_parameters)
- expected exactly once, not yet invoked: ReportMailer.report("this@email.cz", "report.gz", "result")
satisfied expectations:
- expected exactly once, invoked once: #<AnyInstance:ReportComposer>.render(any_parameters)
```

lzap commented:
The code looks good to me, let's put implementation detail of the smart-proxy plugin part aside for this conversation. @ekohl how do you feel about this External IPAM Smart Proxy API? I don't want you to read whole conversation or do careful review as I am planning to test the patch myself, but I would like to hear another opinion about the API we are building here.

Now, failing local tests is something we sometimes see. Try to rebase on top of the fresh develop branch, also do seed and migrate of the database (test environment) that usually helps. What's key is that tests must be passing on our Jenkins so can you start working on that? Just rebase and force-push first to see if that helps. The error "RuntimeError: unknown RestClient::Response.create version (2.1.0, arity -4)" was reported to be issue fixed in 42cf34e62721cc96bd97ca0058d5ac511fd35527.

How I am going to test this: instead installing phpIPAM I am going to write a test plugin following the API that will serve as a dummy IPAM and also a check that the API was designed well. Sorry about the huge delay, I should be much quicker from now on (vacation).

grizzthedj commented:
Sorry @lzap - I have been bogged down with other projects and fires, but will find some time next week to look into resolving the broken tests

ekohl commented on app/models/concerns/orchestration/external_ipam.rb:
```diff
@@ -0,0 +1,79 @@
+    if ip_available
+      subnet.external_ipam_proxy.add_ip_to_subnet(ip, subnet)
+    else
+      self.errors.add :ip, _('IP address ' + ip + ' has already been reserved in External IPAM')
```
[translation](https://projects.theforeman.org/projects/foreman/wiki/translating) needs interpolation because the input to `_()` must be a static string.

ekohl commented on app/models/concerns/orchestration/external_ipam.rb:
```diff
@@ -0,0 +1,79 @@
+  end
+
+  def ip_available
+    subnet.external_ipam_proxy.ip_exists(ip, subnet)['exists'] == false
```
Is `ip_exists` guaranteed to return a hash? Perhaps it'd be better to introduce a `subnet.external_ipam_proxy.ip_exists?(ip, subnet)` method.

ekohl commented on app/models/concerns/orchestration/external_ipam.rb:
```diff
@@ -0,0 +1,79 @@
+    !old.ip.nil? && !old.subnet_id.nil?
+  end
+
+  def ip_available
```
Since it's a boolean, would `ip_available?` be a better name?

ekohl commented on app/helpers/subnets_helper.rb:
```diff
@@ -33,4 +33,8 @@ def subnet_ipam_modes(type)
   end
+
+  def external_ipam?(subnet)
+    subnet&.ipam&.downcase == "external ipam"
```
Later you're repeating the `external ipam` string. Perhaps a `subnet&.ipam&.external?` helper would be useful?

ekohl commented on app/models/subnet.rb:
```diff
@@ -218,6 +233,25 @@ def unused_ip(mac = nil, excluded_ips = [])
+  def validate_against_external_ipam
+    return unless self.errors.full_messages.empty?
+
+    if self.ipam&.downcase == "external ipam"
```
Here the `self.ipam&.external?` helper would also be useful again.

ekohl commented on app/models/subnet.rb:
```diff
@@ -218,6 +233,25 @@ def unused_ip(mac = nil, excluded_ips = [])
+  end
+
+  def subnet_exists_in_external_ipam
+    subnet = JSON.parse(self.external_ipam_proxy.get_subnet(self)) if self.external_ipam_proxy
```
I'd expect the `get_subnet` helper to do the JSON parsing since that interacts with the external service. At this level you shouldn't need to worry about those transport layer encoding.

ekohl commented on app/services/ipam/external_ipam.rb:
```diff
@@ -0,0 +1,33 @@
+        return nil
+      end
+
+      cidr = @subnet.network + '/' + @subnet.cidr.to_s
```
There's already a helper for this:
```suggestion
      cidr = @subnet.network_address
```

ekohl commented on app/views/subnets/_fields.html.erb:
```diff
@@ -8,13 +8,14 @@
                                    "<li><strong>Internal DB</strong> - use internal DB to auto-suggest free IP based on other interfaces on same subnet respecting range if specified, useful mainly with static boot mode <em>(IPv4, IPv6)</em>, preserves natural ordering</li>" +
                                    "<li><strong>Random DB</strong> - same as Internal DB but randomizes results to prevent race conditions <em>(IPv4)</em></li>" +
                                    "<li><strong>EUI-64</strong> - will assign the IPv6 address based on the MAC address of the interface <em>(IPv6)</em></li>" +
+                                   "<li><strong>External IPAM</strong> - will auto-suggest the next available address via an external IPAM Smart-proxy plugin (IPv4)</li>" +
```
Is there a reason to limit external IPAM to IPv4?

ekohl commented on lib/proxy_api/external_ipam.rb:
```diff
@@ -0,0 +1,142 @@
+    #   Response if error:
+    #     {"error":"The specified subnet does not exist in external ipam."}
+    def next_ip(subnet, mac)
+      cidr = subnet.network + '/' + subnet.cidr.to_s if subnet.network.present?
```
When would `subnet.network` not be present? If it isn't, the `next_ip` call will fail and should rather raise an exception. You can also use the helper
```suggestion
      cidr = subnet.network_address if subnet.network.present?
```

ekohl commented on lib/proxy_api/external_ipam.rb:
```diff
@@ -0,0 +1,142 @@
+    #   Response if error:
+    #     {"error":"The specified subnet does not exist in external ipam."}
+    def add_ip_to_subnet(ip, subnet)
+      cidr = subnet.network + '/' + subnet.cidr.to_s if subnet.network.present?
```
Same as in `next_ip`.

ekohl commented on lib/proxy_api/external_ipam.rb:
```diff
@@ -0,0 +1,142 @@
+    #     {"error":"The specified subnet does not exist in external ipam."}
+    def next_ip(subnet, mac)
+      cidr = subnet.network + '/' + subnet.cidr.to_s if subnet.network.present?
+      get("/next_ip?cidr=#{cidr}&mac=#{mac}")
```
Does `get()` accept a hash of params to automatically construct those in the URL? It should also take care of escaping.

grizzthedj replied:
No longer using this workaround `get()` method

ekohl commented on lib/proxy_api/external_ipam.rb:
```diff
@@ -0,0 +1,142 @@
+
+    # TODO: Use ProxyAPI::Resource.parse instead for GET(for consistency), once it has been
+    #       refactored to handle error messages better
+    def get(path, body = nil)
```
Get with a body is very unusual and you're not even using the parameter. It's much more common to have GET parameters so I'd accept a hash of those and add it to the URL.

ekohl commented on lib/proxy_api/external_ipam.rb:
```diff
@@ -0,0 +1,142 @@
+
+    private
+
+    # TODO: Use ProxyAPI::Resource.parse instead for GET(for consistency), once it has been
```
This is IMHO a blocker for this PR. You're not setting up certificates so there can't be authentication. You also miss all the telemetry that it gives you.

I will agree that parsing isn't great, but the class has split HTTP operations and parsing so you can actually just use `get` and manually handle the parsing. You're also not doing any error handling (HTTP 404, 500 etc).

ekohl commented on lib/proxy_api/external_ipam.rb:
```diff
@@ -0,0 +1,142 @@
+    # Get a list of subnets for given external ipam section/group
+    #
+    # Input: group_name(string). The name of the external ipam section/group
+    # Returns: Array of subnets(as json) on success, hash with error otherwise
```
Why return JSON? That's something for the transport between the Smart Proxy and Foreman but why should the rest of Foreman care about this? It could even return a list of `IPAddr` objects. I'd also expect an error to be raised as an exception. Otherwise you need to do error handling manually all over the application which gets messy.

ekohl reviewed (changes requested):
I gave it a quick read through. Most of my concerns are around `lib/proxy_api/external_imap.rb`. My biggest concern in the past with that has been a lack of documentation.

For example, the `next_ip` method: does it claim the IP in the external IPAM or is it just a suggestion? In other words: if I call `next_ip` 255 times in a /24, does that reserve the entire block or are there possible collisions. I don't know the answer and haven't thought about the concrete implementation but these are good things to document. If it doesn't reserve it, you do risk race conditions for example.

The concept of sections also sounds like something unfamiliar to me. Is it something implementation specific?

@lzap also mentioned the use of IDs but that sounds messy. It's implementation specific. I'd prefer that the Smart Proxy implementation would handle this transparently for me. I just give it a subnet and it knows what I'm talking about.

lzap commented:
Thanks for detailed review, I agree with all points and you raised important thing about behavior of API calls - they must be well defined. Let's document them here so we can check everything during the re-review process. Good work, keep it up! I would very much like to see this in Foreman.

ekohl commented:
> I would very much like to see this in Foreman.

While I didn't say it explicitly, I do want to emphasize that this does make sense. My comments are only intended to make it a mature and stable implementation. Thanks for putting in the effort!

grizzthedj commented:
Thanks @ekohl for the feedback. It all makes sense. Will get going on these changes over the next few days.

lzap commented:
One of our customers expressed need for Infoblox IPAM, so once this is ready, I will try to find time to create a competitive Infoblox implementation of the API to test if well the API is designed. Looking forward more work! Thanks guys, this is really a nice feature.

grizzthedj committed:
Some refactoring, removing workaround for smart proxy plugin GET and POST calls

grizzthedj commented:
Hi @lzap @ekohl: Sorry for the delay here.

I have refactored based on the recent feedback, added some documentation, and removed the workaround GET method that I was using. I have also refactored the IPAM proxy plugin to accommodate these changes. 

There are a couple of points/questions that I need to ask to get further feedback on: 

1. **IPv6:** My intention was to handle this in another PR after the first merge happens. That being said, if you think it makes sense to add this functionality now, I would be happy to do so.

2. **Multi-tenant Environments:** In phpIPAM, there is a setting called "Require unique subnets". When this setting is enabled, subnets must be unique, and cannot overlap, across the entire phpIPAM instance. The current implementations of the IPAM proxy plugin and Foreman patch handles this scenario nicely.

If however this setting is disabled, that means duplicate subnets and IP addresses can exist in multiple sections(a section is just a logical group of subnets and IP's). For this scenario, the current implementation is hardcoded to always use the first subnet(index 0), so it still needs some work. 

We need to store the External IPAM section name on the Foreman side so that we can associate Foreman subnets to phpIPAM sections, so that we can handle the overlap and duplication. I think adding a string column to the subnets model should solve this. Perhaps we can name it like "tenant" and not specific to "section", to keep things generic? Although I haven't really looked into it yet, I suspect other IPAM providers(i.e. Infobox, Netbox etc.) have a similar construct.

In terms of handing this in the Foreman UI/UX, when a user creates an External IPAM subnet in Foreman, it would be assumed that "Require unique subnets" is enabled(as a default). In the case where "Require unique subnets" is disabled, users could edit the Subnet in Foreman, and they would be presented with a drop down list of sections, or tenants, whatever is decided from a naming perspective, and they could associate the section/tenant with the foreman Subnet. 

3. **Satellite:** Once this PR gets merged, would this functionality also be available in Red Hat Satellite? If so, how frequent are the release cycles? When could we expect to see this feature in Satellite?

I also recorded a demo video that shows the Foreman/phpIPAM integration in action, so you can get a sense for how things are currently working: 

https://youtu.be/QM5cEXicp9Y

As far as the tests go, I have merged the latest upstream develop to my local branch and am seeing a number of errors and failures. I will try to dig into this while I wait for feedback, then squash the commits etc.
 
Looking forward to your feedback! 

Chris

grizzthedj commented:
Closed and reopened by accident!

grizzthedj committed:
Fixing rubocop errors

pescobar commented:
is this requirement mandatory? `A phpIPAM instance that the Forman Smart Proxy has network connectivity to`

This means that in a distributed setup (e.g. one foreman master and multiple foreman proxies in different data centers) you would need that every proxy has access to the phpIPAM api, right? would it work if only the foreman master has access to the phpIPAM api?

grizzthedj commented:
@pescobar: No. The Foreman master does not talk to the phpIPAM API directly. It does so through the Smart Proxy IPAM Plugin. At least one Smart Proxy in your setup needs to have connectivity to a phpIPAM API for this integration to work. Whichever Smart Proxies(either one, or all of them) have the IPAM feature enabled can then be associated with corresponding subnets in Foreman.

pescobar commented:
@grizzthedj thanks for the clarification

grizzthedj commented:
Closing this PR as there are some issues with the branch. A new PR with a new branch will be submitted.

Here is the link to the new PR: https://github.com/theforeman/foreman/pull/7113
